src/main/java/io/cryostat/net/web/http/api/v2/AbstractV2RequestHandler.java

```diff
@@ -110,8 +111,17 @@ public final void handle(RoutingContext ctx) {
                             || cause instanceof AuthorizationErrorException
                             || cause instanceof KubernetesClientException) {
                         throw new ApiException(401, "HTTP Authorization Failure", ee);
+                    } else if (cause instanceof EmptyOpenShiftTokenException) {
+                        ctx.response()
+                                .putHeader(
+                                        "X-Location",
+                                        OpenShiftAuthManager.getAuthorizationEndpoint())
+                                .putHeader("access-control-expose-headers", "Location")
+                                .setStatusCode(302)
+                                .end();
```

Comment:
I don't have a solid overall design suggestion for exactly how to do this, but here are some unorganized thoughts.

1. This is a common enough case that we should use normal control flow here, not relying on thrown/caught exceptions to control the flow
2. Expanding on that, maybe it makes sense to add a method to the `AuthManager` interface that looks something like `Optional<Headers> suppliedHeadersSufficient(Headers requestHeaders)` (this name sucks). Given the headers provided by the client on the request, it returns an empty Optional if the headers are "sufficient" from the AuthManager's perspective - ex. there is an Authorization header and, for the OpenShiftAuthManager in particular, the auth scheme is Bearer and the token is not blank. If the headers are not sufficient for any reason, then it returns an Optional containing Headers that should be sent back in a response to the client. In this case we would set the Location header. When the `AbstractV2RequestHandler` receives a request it should call `suppliedHeadersSufficient`. If the returned Optional is empty it proceeds as it usually does now. If it is non-empty then it inspects the returned response Headers and short-circuits into an early response to the client, determining the status code by the contents of the Headers (Location header implies a 302 status code, otherwise a 401, or something like that).
That's a really, really roughed out idea and needs some refinement. I already don't particularly like that the AuthManagers have to know anything about HTTP headers, and I'd like to avoid adding any more of a coupling there, so passing in and out the Headers objects directly isn't ideal and I would rather some abstraction be introduced here to represent `Your auth request is OK | Your auth request might be OK but there's another step required | Your auth request is invalid | Your auth request was rejected`. My rough idea above embeds this information into the Headers object and conditional logic about whether certain keys are present in the map or the form of the associated values.
3. The `authorizationUrl` shouldn't just be a "raw" `public static` field - it should only be accessible by calling a method on the `OpenShiftAuthManager`, which can have some kind of memoization so that it it isn't performing an HTTP request every time some caller needs to know the OAuth endpoint URL. You might just be able to reuse the `Future` returned by your existing `webClient` request for that - calling `.get()` on the response should just return back the value immediately on the 2nd, 3rd, nth invocation, I think. If not then it's fairly straightforward to memoize it with `CompletableFuture.complete`/`CompletableFuture.supplyAsync` and return that reference every time some caller needs it.
4. > except I'm not sure how to reach the AuthPostHandler if validateRequestAuthorization() throws a 401 before writeResponse(ctx, handle(requestParams)) is called
If the `AbstractV2RequestHandler`'s assumptions don't work or don't make sense for your concrete implementation class, then don't extend `AbstractV2RequestHandler` :-) . You can rename `AbstractV2RequestHandler` to `AbstractAuthenticatedV2RequestHandler` and create another one that doesn't do authorization checks on its own, and pull the authorization logic you do need to implement into your concrete `AuthPostHandler` instead.

Reply:
My latest refactor removes the headers inside `AuthPostHandler` and removed the custom exception. I have one last question before I add tests and open the PR:

The web browser will make an automatic GET request whenever it sees a response header named `Location` with a status code `302` and won't follow the redirect url. Some possible solutions, including setting the status code to a `401` or `403` instead of a `302` are outlined here: https://stackoverflow.com/a/52290824 

I don't think an error status code would be appropriate here, so I renamed the `Location` header to `X-Location`. What do you think?

Reply:
Sure, that makes sense to me.